docs/design/RUC-on-type.md

```diff
@@ -0,0 +1,189 @@
+
+## RequiresUnreferencedCode on a type warns on instance constructor or static method access
+Having RequiresUnreferencedCode on the type implies that any instance constructor or static method is considered dangerous for trimming. Following are examples of this behavior:
+
+```C#
+[RequiresUnreferencedCode ("Message for --ClassWithRequiresUnreferencedCode--")]
+class ClassWithRequiresUnreferencedCode
+{
+    public static void StaticMethod () { }
+
+    public void NonStaticMethod () { }
+}
+
+static void TestRequiresUnreferencedCodeInClass ()
+{
+    // IL2026: TestRequiresUnreferencedCodeInClass(): Using method 'ClassWithRequiresUnreferencedCode.StaticMethod()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Message for --ClassWithRequiresUnreferencedCode--.
+    ClassWithRequiresUnreferencedCode.StaticMethod ();
+    // IL2026: TestRequiresUnreferencedCodeInClass(): Using method 'ClassWithRequiresUnreferencedCode.ClassWithRequiresUnreferencedCode()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Message for --ClassWithRequiresUnreferencedCode--.
+    var classWithRequiresUnreferencedCode = new ClassWithRequiresUnreferencedCode ();
+
+    ClassWithRequiresUnreferencedCode.NonStaticMethod();
+}
+```
+With the previous example we can see that calling the static method or generating a new instance of the type will result in generating a warning. Also, in the case of the instance method 'NonStaticMethod()' we dont generate a warning since we consider that in order to call this method an instance of the type needs to be generated first causing the constructor to warn with IL2026.
+
+## RequiresUnreferencedCode on a type suppresing other warnings
+The behavior of RequiresUnreferencedCode suppressions on a method is that it will suppress the warnings inside the method body and only generate a single IL2026 warning instead.
+```C#
+[RequiresUnreferencedCode ("This method calls dangeours code")]
+public static void MethodWithDangerousCode () 
+{
+    // Dangerous code that generetares a trimming warning
+
+static void TestSuppressionOnMethod ()
+{
+    // IL2026: TestSuppressionOnMethod(): Using method 'MethodWithDangerousCode()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. This method calls dangeours code.
+    MethodWithDangerousCode()
+}
+```
+
+Similarly RequiresUnreferencedCode on a type suppresses other warnings, with the difference that it will silence warnings from all methods in the type and attributes (except maybe static constructors, pending discussion)
```

Comment:
I have mixed feelings about suppressing warnings from attributes. Assume `Problematic` is a problematic attribute. Maybe because the attribute itself is marked as RUC.

```csharp
[RUC]
[Problematic]
class Foo { }
```

Now I do: `typeof(Foo).GetCustomAttribute(ProblematicAttribute)`. Where do I get a warning that we just executed trim-unfriendly code?

There will be similar problems with custom attributes on instance methods; static methods will be covered because we first need to reflect on the method itself and that's a warning if the method is static; maybe we could warn on reflection-accessing instance methods of RUC types and it will be fixed there too?.